packages/flutter/lib/src/widgets/basic.dart

```diff
@@ -2369,42 +2476,138 @@ class UnconstrainedBox extends SingleChildRenderObjectWidget {
   ///  * [AlignmentDirectional] for [Directionality]-aware alignments.
   final AlignmentGeometry alignment;
 
-  /// The axis to retain constraints on, if any.
+  /// {@template flutter.widgets.constraintsTransform}
+  /// The function used to transform the incoming [BoxConstraints], to size
+  /// [child].
   ///
-  /// If not set, or set to null (the default), neither axis will retain its
-  /// constraints. If set to [Axis.vertical], then vertical constraints will
-  /// be retained, and if set to [Axis.horizontal], then horizontal constraints
-  /// will be retained.
-  final Axis? constrainedAxis;
+  /// The function must return a [BoxConstraints] that is
+  /// [BoxConstraints.isNormalized].
+  ///
+  /// See [ConstraintsTransformBox] for predefined common
+  /// [BoxConstraintsTransform]s.
+  /// {@endtemplate}
+  final BoxConstraintsTransform constraintsTransform;
 
   /// {@macro flutter.material.Material.clipBehavior}
   ///
   /// Defaults to [Clip.none].
   final Clip clipBehavior;
 
+  final String _debugTransformLabel;
+
   @override
-  void updateRenderObject(BuildContext context, covariant RenderUnconstrainedBox renderObject) {
+  RenderConstraintsTransformBox createRenderObject(BuildContext context) {
+    return RenderConstraintsTransformBox(
+      textDirection: textDirection ?? Directionality.maybeOf(context),
+      alignment: alignment,
+      constraintsTransform: constraintsTransform,
+      clipBehavior: clipBehavior,
+    );
+  }
+
+  @override
+  void updateRenderObject(BuildContext context, covariant RenderConstraintsTransformBox renderObject) {
     renderObject
       ..textDirection = textDirection ?? Directionality.maybeOf(context)
+      ..constraintsTransform = constraintsTransform
       ..alignment = alignment
-      ..constrainedAxis = constrainedAxis
       ..clipBehavior = clipBehavior;
   }
 
   @override
-  RenderUnconstrainedBox createRenderObject(BuildContext context) => RenderUnconstrainedBox(
-    textDirection: textDirection ?? Directionality.maybeOf(context),
-    alignment: alignment,
-    constrainedAxis: constrainedAxis,
-    clipBehavior: clipBehavior,
-  );
+  void debugFillProperties(DiagnosticPropertiesBuilder properties) {
+    super.debugFillProperties(properties);
+    properties.add(DiagnosticsProperty<AlignmentGeometry>('alignment', alignment));
+    properties.add(EnumProperty<TextDirection>('textDirection', textDirection, defaultValue: null));
+
+    final String? debugTransformLabel = _debugTransformLabel.isNotEmpty
+      ? _debugTransformLabel
+      : _knownTransforms[constraintsTransform];
```

Comment:
nit: name this _debugKnownTransforms to message that it's only supposed to be used within debug* methods?